TypeCobol/Analysis/AnalyzerProvider.cs

```diff
@@ -10,103 +12,177 @@ namespace TypeCobol.Analysis
+                .Select(qaActivator => SecureCreateAnalyzer(() => qaActivator(options)))
+                .Where(qaAnalyzer => qaAnalyzer != null);
+
+            return MergeAnalyzers(qaFromProviders, qaFromActivators);
         }
 
-        public override ISyntaxDrivenAnalyzer[] CreateSyntaxDrivenAnalyzers(TypeCobolOptions options, TextSourceInfo textSourceInfo)
+        public ISyntaxDrivenAnalyzer[] CreateSyntaxDrivenAnalyzers([NotNull] TypeCobolOptions options, [NotNull] TextSourceInfo textSourceInfo)
         {
-            var fromBase = base.CreateSyntaxDrivenAnalyzers(options, textSourceInfo);
-            return Concat(fromBase, p => p.CreateSyntaxDrivenAnalyzers(options, textSourceInfo));
+            var sdaFromProviders = SecureCreateAnalyzersFromProviders(analyzerProvider => analyzerProvider.CreateSyntaxDrivenAnalyzers(options, textSourceInfo));
+
+            var sdaFromActivators = _sdaActivators?
+                .Select(sdaActivator => SecureCreateAnalyzer(() => sdaActivator(options, textSourceInfo)))
+                .Where(sdaAnalyzer => sdaAnalyzer != null);
+
+            return MergeAnalyzers(sdaFromProviders, sdaFromActivators);
         }
 
-        public override IQualityAnalyzer[] CreateQualityAnalyzers(TypeCobolOptions options)
+        /// <summary>
+        /// Add an activation delegate to produce a new instance of ISyntaxDrivenAnalyzer.
+        /// </summary>
+        /// <param name="sdaActivator">Non-null Func delegate to create a new ISyntaxDrivenAnalyzer.</param>
+        public void AddActivator([NotNull] Func<TypeCobolOptions, TextSourceInfo, ISyntaxDrivenAnalyzer> sdaActivator)
         {
-            var fromBase = base.CreateQualityAnalyzers(options);
-            return Concat(fromBase, p => p.CreateQualityAnalyzers(options));
+            if (_sdaActivators == null)
+            {
+                _sdaActivators = new List<Func<TypeCobolOptions, TextSourceInfo, ISyntaxDrivenAnalyzer>>();
+            }
+            _sdaActivators.Add(sdaActivator);
+        }
+
+        /// <summary>
+        /// Add an activation delegate to produce a new instance of IQualityAnalyzer.
+        /// </summary>
+        /// <param name="qaActivator">Non-null Func delegate to create a new IQualityAnalyzer.</param>
+        public void AddActivator([NotNull] Func<TypeCobolOptions, IQualityAnalyzer> qaActivator)
+        {
+            if (_qaActivators == null)
+            {
+                _qaActivators = new List<Func<TypeCobolOptions, IQualityAnalyzer>>();
+            }
+            _qaActivators.Add(qaActivator);
         }
 
         /// <summary>
         /// Add an analyzer provider to this one.
         /// </summary>
         /// <param name="provider">Instance of IAnalyzerProvider.</param>
-        public void AddProvider(IAnalyzerProvider provider)
+        public void AddProvider([NotNull] IAnalyzerProvider provider)
         {
+            Debug.Assert(provider != null);
```

Comment:
Either remove this assertion or add it for the 2 other AddActivator methods to improve consistency.

Reply:
Assertion removed as the methods have the `[NonNull]` annotation.